core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java

```diff
@@ -134,6 +141,13 @@ class AddQuadOperation implements Operation {
 
 		@Override
 		public void execute() throws IOException {
+			if (!unusedIds.isEmpty()) {
+				// these ids are used again
+				unusedIds.remove(s);
+				unusedIds.remove(p);
+				unusedIds.remove(o);
+				unusedIds.remove(c);
+			}
 			boolean wasNew = tripleStore.storeTriple(s, p, o, c, explicit);
```

Comment:
I'm wondering about how concurrency is handled. Is it possible that a concurrent operation can be adding/removing items from `unusedIds`? Could it be possible that `unusedIds` is empty on line 144, but due to a concurrent operation s, p, o or c could be added to `unusedIds` and removed from the value store right before the triple is stored on line 151?

Reply:
As LmdbStore is based on SnapshotSailStore there shouldn't be any concurrency issues. Every write access is executed only by one thread at any point in time.

Reply:
It's a bit hard to see. This code is called asynchronously from startTransaction(), but also from addStatement() (LmdbSailStore line 673).

Reply:
Yes. This depends on the mode of the LMDB store. If multi-threading is active then it is executed via a separate worker thread for the triple store, else it is directly executed synchronously. Both modes are never mixed.